proposals/0173-sbpf-instruction-encoding-improvements.md

```diff
@@ -0,0 +1,98 @@
+---
+simd: '0173'
+title: SBPF instruction encoding improvements
+authors:
+  - Alexander Meißner
+category: Standard
+type: Core
+status: Draft
+created: 2024-09-05
+feature: TBD
+extends: SIMD-0161
+---
+
+## Summary
+
+There are some instruction with questionable encodings, that when slightly
+adjusted, could significantly simplify verification and execution of programs.
+
+## Motivation
+
+The instruction `lddw dst, imm` is currently the only instruction which takes
+two instruction slots. This proposal splits it into a two one-slot instruction
+sequence: `mov32 dst, imm` and an introduced `hor64 dst, imm`. This way all
+instructions will be exactly one slot long which will simplyfiy:
+
+- Calculating the number of instructons in a program will no longer require a
+full linear scan. A division of the length of the text section by the
+instruction slot size will suffice.
+- The instruction meter will no longer have to skip one instruction slot when
+counting a `LDDW` instruction.
+- Jump and call instructions will no longer have to verify that the desination
+is not the second half of a `LDDW` instruction.
+- The verifier will no longer have to check that `LDDW` instructions are
+complete and its first or second half does not occur without the other on its
+own.
```

Comment:
Removing LDDW does not eliminate relocations. When the program contains an address load with the new instruction pair `mov32/hor64`, we may still need a relocation. In this case, I believe we should check (at least during the relocation process) we are writing the relocated address to a `mov32/hor64` pair instead of overwriting another instruction.

Reply:
That is a problem of the toolchain. In general we should aim to apply all relocations at link time.